src/net/sourceforge/kolmafia/request/RelayRequest.java

```diff
@@ -3500,6 +3514,156 @@ public boolean pollForCompletion(String type) {
+   */
+  private void handleJsonApi(String request) {
+    this.contentType = "application/json";
+    JSONObject json;
+    try {
+      json = JSON.parseObject(request);
+    } catch (JSONException e) {
+      jsonError("Invalid JSON object in request.");
+      return;
+    }
+
+    var result = new JSONObject();
+
+    var properties = json.get("properties");
+    if (properties instanceof JSONArray propertiesArray) {
+      var nonString = propertiesArray.stream().filter(name -> !(name instanceof String)).toList();
+      if (!nonString.isEmpty()) {
+        jsonError("Invalid property names " + JSON.toJSONString(nonString));
+        return;
+      }
+      result.put(
+          "properties",
+          ((JSONArray) properties)
+              .stream()
+                  .map(name -> name instanceof String ? Preferences.getString((String) name) : null)
+                  .toList());
+    }
+
+    var functions = json.get("functions");
+    if (functions instanceof JSONArray functionsArray) {
+      var nonMatching =
+          functionsArray.stream()
+              .filter(
+                  func -> {
+                    if (!(func instanceof JSONObject)) return true;
+                    var funcObject = (JSONObject) func;
+                    var name = funcObject.get("name");
+                    var args = funcObject.get("args");
+                    if (!(name instanceof String) || !(args instanceof JSONArray)) return true;
+                    return ((JSONArray) args).contains(null);
+                  })
+              .toList();
+      if (!nonMatching.isEmpty()) {
+        jsonError("Invalid function calls " + JSON.toJSONString(nonMatching));
+        return;
+      }
+
+      // Everything validated. Transform all function arguments.
+      var functionsResult = new JSONArray(functionsArray.size());
+      for (var func : (JSONArray) functions) {
+        var funcObject = (JSONObject) func;
+        var name = (String) funcObject.get("name");
+        var args = (JSONArray) funcObject.get("args");
+        var transformedArguments = args.stream().map(this::transformApiArgument).toList();
+        var badIndices =
+            IntStream.range(0, transformedArguments.size())
+                .filter(i -> transformedArguments.get(i) == null)
+                .toArray();
+        if (badIndices.length > 0) {
+          jsonError(
+              "Invalid arguments to "
+                  + name
+                  + ": "
+                  + JSON.toJSONString(Arrays.stream(badIndices).mapToObj(args::get)));
+          return;
+        }
+
+        if (name.equals("identity") && args.size() == 1) {
+          functionsResult.add(JSONValueConverter.asJSON(transformedArguments.get(0).asProxy()));
+          continue;
+        }
+
+        Class<?>[] argClasses = new Class[args.size() + 1];
+
+        argClasses[0] = ScriptRuntime.class;
+        Arrays.fill(argClasses, 1, args.size() + 1, Value.class);
```

Comment:
I figure this was copied from `LibraryFunction` -- might be worth extracting into its own little function as the logic needs to be the same here and there